File: TraversalDemo/src/main/scala/Interpreter.scala
// Pattern-matching interpreter
object Interpreter {
  type SymTab = scala.collection.mutable.Map[String, Value]

  def interpret(p: Program): Unit = {
    val symtab = scala.collection.mutable.Map[String, Value]()

    for (stmt <- p.stmts) {
      interpret(stmt, symtab)
    }
  }

  def interpret(stmt: Statement, symtab: SymTab): Unit = stmt match {
    case Decl(id, INT) =>
      symtab(id.lexeme) = IntValue(0)

    case Decl(id, REAL) =>
      symtab(id.lexeme) = RealValue(0.0)

    case Decl(id, UNKNOWN) =>
      // This shouldn't happen

    case Assign(lhs, rhs) => 
      symtab(lhs.lexeme) = interpret(rhs, symtab)

    case PrintInt(expr) =>
      println(interpret(expr, symtab))

    case PrintReal(expr) => 
      println(interpret(expr, symtab))
  }

  def interpret(expr: Expression, symtab: SymTab): Value = expr match {
    case Add(left, right) => {
      val a = interpret(left, symtab)
      val b = interpret(right, symtab)
      
      (a, b) match {
        case (IntValue(m), IntValue(n)) => IntValue(m + n)
        case (RealValue(x), RealValue(y)) => RealValue(x + y)
        case _ => sys.error("Type mismatch")
      }
    }

    case Num(lexeme) =>
      if (lexeme.contains('.')) {
        RealValue(lexeme.toDouble)
      } else {
        IntValue(lexeme.toInt)
      }

    case Var(lexeme) =>
      if (symtab.isDefinedAt(lexeme)) {
        symtab(lexeme)
      } else {
        sys.error(s"Unknown variable $lexeme")
      }
  }
}

File: TraversalDemo/src/main/java/edu/depauw/demo/visitor/Interpreter.java
package edu.depauw.demo.visitor;

import java.util.HashMap;
import java.util.Map;

import edu.depauw.demo.common.IntValue;
import edu.depauw.demo.common.RealValue;
import edu.depauw.demo.common.Value;
import edu.depauw.demo.visitor.ast.*;

public class Interpreter implements StatementVisitor<Void, Void>, ExpressionVisitor<Void, Value> {
	private Map<String, Value> symtab;
	
	public Interpreter() {
		symtab = new HashMap<>();
	}
	
	public static void interpret(Program program) {
		Interpreter interpreter = new Interpreter();
		
		for (Statement stmt : program.getStatements()) {
			stmt.accept(interpreter, null);
		}
	}

	@Override
	public Void visit(Decl decl, Void t) {
		String lexeme = decl.getId().getLexeme();
		Type type = decl.getType();
		
		if (type == Type.INT) {
			symtab.put(lexeme, new IntValue(0));
		} else {
			symtab.put(lexeme, new RealValue(0.0));
		}
		
		return null;
	}

	@Override
	public Void visit(Assign assign, Void t) {
		String lexeme = assign.getLhs().getLexeme();
		Expression rhs = assign.getRhs();
		
		symtab.put(lexeme, rhs.accept(this, t));
		return null;
	}

	@Override
	public Void visit(PrintInt printInt, Void t) {
		Expression expr = printInt.getExpr();

		System.out.println(expr.accept(this, t));
		return null;
	}

	@Override
	public Void visit(PrintReal printReal, Void t) {
		Expression expr = printReal.getExpr();
		
		System.out.println(expr.accept(this, t));
		return null;
	}

	@Override
	public Value visit(Add add, Void t) {
		Value a = add.getLeft().accept(this, t);
		Value b = add.getRight().accept(this, t);
		
		if (a instanceof IntValue && b instanceof IntValue) {
			IntValue aInt = (IntValue) a;
			IntValue bInt = (IntValue) b;

			return new IntValue(aInt.getValue() + bInt.getValue());
		} else if (a instanceof RealValue && b instanceof RealValue) {
			RealValue aReal = (RealValue) a;
			RealValue bReal = (RealValue) b;

			return new RealValue(aReal.getValue() + bReal.getValue());
		} else {
			throw new RuntimeException("Type mismatch");
		}
	}

	@Override
	public Value visit(Num num, Void t) {
		String lexeme = num.getLexeme();

		if (lexeme.contains(".")) {
			return new RealValue(Double.parseDouble(lexeme));
		} else {
			return new IntValue(Integer.parseInt(lexeme));
		}
	}

	@Override
	public Value visit(Var var, Void t) {
		String lexeme = var.getLexeme();

		if (symtab.containsKey(lexeme)) {
			return symtab.get(lexeme);
		} else {
			throw new RuntimeException("Unknown variable " + lexeme);
		}
	}
}
